src/mongocxx/exception/error_code.hpp

```diff
@@ -31,6 +31,92 @@ enum class error_code : std::int32_t {
     k_unknown_write_concern,
 };
 
+MONGOCXX_API const std::error_code& bson_invalid_error();
```

Comment:
I realized what was bugging me about this over the weekend. While this does make us independent of the numeric values of the libbson/libmongoc error codes, it is still exposing the C level error schema. The difference is that it is symbolic rather than numeric. That is a good step, but it still feels like a leaky abstraction.

We also have the issue raised in https://jira.mongodb.org/browse/CXX-808

This suggests to me that we haven't quite arrived where we want to be as far as the error abstraction. Lets try to find some time to talk it through. @hanumantmk @ajdavis @amidvidy @samantharitter - I'd be interested in all of your thoughts on this issue, as we don't have a lot of time to resolve it if we want to make our GA date.

Overall we have so far done a very good job making our use of the C driver completely opaque, but this is definitely a troublesome area.